File: Competitive Programming 3/Section 1.4/Palidromes/401 - Palindromes/Python/main.py
from sys import stdin
from copy import copy

ans=[("is not a palindrome","is a mirrored string"),("is a regular palindrome","is a mirrored palindrome")]
reverse={
    'A':'A',
    'E':'3',
    'H':'H',
    'I':'I',
    'J':'L',
    'L':'J',
    'M':'M',
    'O':'O',
    'S':'2',
    'T':'T',
    'U':'U',
    'V':'V',
    'W':'W',
    'X':'X',
    'Y':'Y',
    'Z':'5',
    '1':'1',
    '2':'S',
    '3':'E',
    '5':'Z',
    '8':'8'
}
def getPalindromeValue(s):
    lim = len(s) >> 1
    for i in range(lim):
        if s[i]!=s[len(s)-1-i]:
            return 0
    return 1
def getMirroredString(s):
    sChars = list(s)
    for i in range(len(sChars)):
        if s[i] in reverse:
            sChars[i]=reverse[s[i]]
        else:
            return None
    return ''.join(sChars)
def getMirroredPalindromeValue(s):
    mirroredString = getMirroredString(s)
    if not mirroredString:
        return 0
    #compared if mirroredString backwards is the same as s
    for i in range(len(s)):
        if s[i]!=mirroredString[len(s)-1-i]:
            return 0
    return 1
def main():
    for s in stdin:
        s=s.rstrip()
        isPalindrome,isMirrored = getPalindromeValue(s),getMirroredPalindromeValue(s)
        print("{} -- {}.\n".format(s,ans[isPalindrome][isMirrored]))
main()

File: Competitive Programming 3/Section 1.4/Palidromes/401 - Palindromes/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

string ans[2][2] = {{"is not a palindrome","is a mirrored string"},{"is a regular palindrome","is a mirrored palindrome"}};
map<char,char> reversed = {
    make_pair('A','A'),
    make_pair('E','3'),
    make_pair('H','H'),
    make_pair('I','I'),
    make_pair('J','L'),
    make_pair('L','J'),
    make_pair('M','M'),
    make_pair('O','O'),
    make_pair('S','2'),
    make_pair('T','T'),
    make_pair('U','U'),
    make_pair('V','V'),
    make_pair('W','W'),
    make_pair('X','X'),
    make_pair('Y','Y'),
    make_pair('Z','5'),
    make_pair('1','1'),
    make_pair('2','S'),
    make_pair('3','E'),
    make_pair('5','Z'),
    make_pair('8','8')
};
int getPalindromeValue(const string & s){
    auto lim = s.size() >> 1;
    for(size_t i=0;i<lim;i++)
        if(s[i]!=s[s.size()-1-i])
            return 0;
    return 1;
}
string getMirroredString(string s){
    for(size_t i=0;i<s.size();i++){
        auto it = reversed.find(s[i]);
        if(it==reversed.end())
            return "";
        else
            s[i]=it->second;
    }
    return s;  
}
int getMirroredPalindromeValue(const string &s){
    string mirroredString = getMirroredString(s);
    if(mirroredString.empty())
        return 0;
    //compared if mirroredString backwards is the same as s
    for(size_t i=0;i<s.size();i++)
        if(s[i]!=mirroredString[s.size()-1-i])
            return 0;
    return 1;
}

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    string s;
    while(cin>>s){
        int isPalindrome = getPalindromeValue(s), isMirrored = getMirroredPalindromeValue(s);
        cout << s << " -- " << ans[isPalindrome][isMirrored] << ".\n\n";
    }
    return 0;
}